services/federator/src/Federator/Options.hs

```diff
@@ -20,11 +20,55 @@
 
 module Federator.Options where
 
+import qualified Control.Lens as Lens
 import Data.Aeson
+import qualified Data.Aeson as Aeson
+import Data.Domain (Domain ())
 import Imports
 import System.Logger.Extended
 import Util.Options
 
+newtype FederationAllowedDomains = FederationAllowedDomains {allowedDomains :: [Domain]}
+  deriving (Eq, Show, Generic)
+  deriving newtype (FromJSON, ToJSON)
+
+defFederationAllowedDomains :: FederationAllowedDomains
+defFederationAllowedDomains = FederationAllowedDomains []
+
+-- FUTUREWORK: Support a DenyList
+data FederationStrategy
+  = -- | This backend allows federating with any other Wire-Server backend
+    WithEveryone
+  | -- | Any backend explicitly configured in a FederationAllowList
+    WithAllowList
+  deriving (Eq, Show, Generic, Bounded, Enum)
+
+instance ToJSON FederationStrategy where
+  toJSON WithEveryone = "with_everyone"
+  toJSON WithAllowList = "with_allow_list"
+
+instance FromJSON FederationStrategy where
+  parseJSON = withText "FederationStrategy" $ \case
+    "with_everyone" -> pure WithEveryone
+    "with_allow_list" -> pure $ WithAllowList
+    _ ->
+      fail $
+        "unexpected value for FederationStrategy settings: "
+          <> "expected one of "
+          <> show (Aeson.encode <$> [(minBound :: FederationStrategy) ..])
+
+-- | Options that persist as runtime settings.
+data RunSettings = RunSettings
+  { -- | Would you like to federate with everyone or only with a select set of other wire-server installations?
+    setFederationStrategy :: !(FederationStrategy),
+    -- | setFederationAllowedDomains only has an effect if 'setFederationStrategy' is with_allow_list
+    -- (defaults to empty list if unset)
+    setFederationAllowedDomains :: !(Maybe FederationAllowedDomains)
```

Comment:
Thinking about this again, I think it comes down to how we want this to behave for someone configuring the service. If you only put `setFederationStrategy: allow_list`, but don't specify an allow-list, should it complain?
It could for example be a typo, which then gets silently ignored and you end up not being able to federate with anyone.

Reply:
Good point. What should the yaml options look like then? While Haskell supports complicated types, in yaml we only have strings and objects.

So

```yaml
federationStrategy: "everyone"
```
or
```yaml
federationStrategy:
  allow_list:
    - example.com
    - ....
```

?

That looks a little weird to me.

Should there be a custom option-validating step at startup time which can take multiple parameters and check them for consistency?